src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryAdder.java

```diff
@@ -256,7 +256,7 @@
                 for (LibraryRecord library : action.getLibraries()) {
                     FilePath libResources = libs.child(library.getDirectoryName() + "/resources/");
                     FilePath f = libResources.child(name);
-                    if (!new File(f.getRemote()).getCanonicalFile().toPath().startsWith(libResources.absolutize().getRemote())) {
+                    if (!new File(f.getRemote()).getCanonicalFile().toPath().startsWith(new File(libResources.getRemote()).getCanonicalPath())) {
```

Comment:
```suggestion
                    if (!new File(f.getRemote()).getCanonicalFile().toPath().startsWith(new File(libResources.getRemote()).getCanonicalFile().toPath())) {
```
?

Reply:
Though even better is to use NIO fully:
```suggestion
                    if (!Paths.get(f.getRemote()).toRealPath().startsWith(Paths.get(libResources.getRemote()).toRealPath())) {
```

Reply:
`Path.toRealPath` throws if the path does not exist on the file system, so we would have to change things a bit more if we wanted to keep attackers from being able to probe the existence of files using symlinks.

Is there a meaningful difference between `Path.startsWith(String)` and `Path.startsWith(Path)`? They both return false for things like `foo/bar` starting with `f`.

Reply:
> Mm, I see. So there is no NIO equivalent to File.getCanonicalPath?

Not directly as far as I know. I did consider using NIO in the original patch and catching `NoSuchFileException` so that error messages would be the same regardless of file existence, but it seemed simpler to use `getCanonicalFile`.

> I do not know.

Ok, I just wasn't sure if it was a style-related suggestion or if I was missing an important detail. As far as I know the current code with `Path.startsWith(f.getCanonicalPath())` and the proposed code with `Path.startsWith(f.getCanonicalFile().toPath())` should have the same behavior.

Reply:
Probably. The weird corner cases are likely to be on Windows. I think the current code is secure and likely to fix the regression for most if not all users.

Reply:
Ok, thanks for elaborating. I'll merge the code as-is and will follow up as needed.